core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -666,12 +670,21 @@ class TimeSeriesShard(val dataset: Dataset,
     indexRb.endRecord(false)
   }
 
+  private def isActivelyIngesting(i: Integer): Boolean = {
+    activelyIngesting.synchronized { activelyIngesting.get(i) }
+  }
+
   // scalastyle:off method.length
   private def doFlushSteps(flushGroup: FlushGroup,
                            partitionIt: Iterator[TimeSeriesPartition]): Task[Response] = {
     val tracer = Kamon.buildSpan("chunk-flush-task-latency-after-retries")
       .withTag("dataset", dataset.name)
       .withTag("shard", shardNum).start()
+
+    // Clean up lingering chunk map shared locks, left over from previous failures. The
+    // makeFlushChunks call below is what creates an iterator which holds a shared lock.
+    ChunkMap.validateNoSharedLocks()
```

Comment:
This call should happen on the thread that starts consuming the chunk iterator which happens in writeChunksFuture below. 

Whereas right now, it is running on the single threaded ingestion scheduler.

The lazy evaluation is quite tricky. I added some comments with "VERY IMPORANT". Please feel free to add more if necessary.

Reply:
I've removed all this stuff for now. It should go into another commit.